data_model/src/expression.rs

```diff
@@ -24,6 +24,150 @@ use serde::{Deserialize, Serialize};
 
 use super::{query::QueryBox, Value, ValueBox};
 
+/// Generate expression structure and basic impls for it.
+///
+/// # Syntax
+///
+/// Basic syntax:
+///
+/// ```ignore
+/// gen_expr_and_impls! {
+///     /// Comment
+///     #[derive(Derives)]
+///     pub Expr(param1: Type1, param2: Type2, param3: Type3, ...) -> OutputType
+/// }
+/// ```
+///
+/// The macro has three syntax forms to specify parameters:
+/// - One unnamed parameter. In that case, the parameter name will be `expression`.
+/// - Two unnamed parameters.
+///   In that case, the parameter names will be `left` and `right` respectively.
+/// - Any number of named parameters.
+///
+/// The macro has two syntax forms to specify result:
+/// - With the actual result type after the arrow (`->`).
+///   In that case, `impl From<$i> for EvaluatesTo<$result_type>` will be generated.
+/// - With `?` sign as a result type.
+///   In that case `impl From<$i> for EvaluatesTo<$result_type>` **won't** be generated.
+///
+/// See the example and further usage for more details.
+///
+/// # Example
+///
+/// ```ignore
+/// gen_expr_and_impls! {
+///     /// Evaluates to the sum of left and right expressions.
+///     #[derive(Debug)]
+///     pub Add(u32, u32) -> u32
+/// }
+///
+/// // Will generate the following code:
+///
+/// /// Evaluates to the sum of left and right expressions.
+/// #[derive(Debug)]
+/// pub struct Add {
+///     #[allow(missing_docs)]
+///     pub left: EvaluatesTo<u32>,
+///     #[allow(missing_docs)]
+///     pub right: EvaluatesTo<u32>,
+/// }
+///
+/// impl Add {
+///     /// Number of underneath expressions
+///     #[inline]
+///     pub fn len(&self) -> usize {
+///         self.left.len() + self.right.len() + 1
+///     }
+///     /// Construct new [`Add`] expression
+///     pub fn new(left: impl Into<EvaluatesTo<u32>>, right: impl Into<EvaluatesTo<u32>>) -> Self {
+///         Self {
+///             left: left.into(),
+///             right: right.into(),
+///         }
+///     }
+/// }
+///
+/// impl From<Add> for ExpressionBox {
+///     fn from(expression: Add) -> Self {
+///         Expression::Add(expression).into()
+///     }
+/// }
+///
+/// impl From<Add> for EvaluatesTo<u32> {
+///     fn from(expression: Add) -> Self {
+///         EvaluatesTo::new_unchecked(expression.into())
+///     }
+/// }
+/// ```
+macro_rules! gen_expr_and_impls {
+    // Case: one unnamed parameter with known result type
```

Comment:
I believe you could decrease number of variants from 6 to 3 + 1 additional internal rule by utilize [`incremental tt-matcher`](https://veykril.github.io/tlborm/decl-macros/patterns/tt-muncher.html).

```suggestion
    // Case: one unnamed parameter
    ($(#[$me:meta])* $v:vis $i:ident($first_type:ty $(,)?) -> $($tt:tt)*) => {
        gen_expr_and_impls!(impl_basic $(#[$me])* $v $i(expression: $first_type));
        gen_expr_and_impls!(impl_extra_convert $i $($tt)*);
    };
```

Reply:
Cool idea!

Reply:
Actually I can also forward cases like *"one unnamed parameter"* and *"two unnamed parameters"* into generic case *"any number of named parameters"*